Fully implement capable enemies, closes #27, closes #37, closes #34

HN67 opened the pull request:
Implement multiple features culminating in the ability to spawn shooting enemies.
This branch derived from mortal, so this pull request renders that one unneeded.
1) Extract Entity into various interfaces (Mortal branch)
2) Implements Gun architecture to allow complex Guns with simple player logic
3) Create WanderPresence "AI"
4) Create ShooterEntity and ShooterSpawner
Implementing ShooterSpawner proved conceptually difficult due to Copyable conflicts, but eventually I just decided to implement AmbientSpawner twice, using a base logic class. (The issue is that technically a Shooter is not a Copyable<Shooter>, even though it acts like one. Attempting to make it inherit Copyable<Shooter> causes a generic collision.)

This will definitely need some gameplay re balancing, since right now the enemies are super scary (and they spawn quickly). Making the shooter enemies not use their gun so often would help.

HN67 committed:
Change Spawner to not directly inherit Interactive

HN67 committed:
Create Gun class

HN67 committed:
Create Mortal interface, simplify Spawner

Most likely in a bugged state, since Entities (i.e. Player) that are also Spawners currently will not be removed from the Spawner list.

HN67 committed:
Switch TextRoom to use multiple sets of API interfaces

HN67 committed:
Merge branch 'mortal' into gun

HN67 committed:
Remove alive method from Gun

HN67 committed:
Integrate Gun architecture into Player

HN67 committed:
Create WanderPresence and add proper Presence copy implementations

HN67 committed:
Create ShooterEntity and make AmbientSpawner techinically spawns Shooters

However, the Shooters produced by the Spawner are not recognized by the Room, only treated as entities.

HN67 committed:
Make Spawne a Generic instead of Entity bound

HN67 committed:
Extract Physical interface

HN67 committed:
Generalize AmbientSpawner and integrate Shooter

Note, currently AmbientSpawner is unable to consume a Shooter, breaking the code. Working on making Shooter recognizable as a Copyable<Shooter>

HN67 committed:
Extract spawning logic into base class, and create shooter spawner

HN67 committed:
Integrate shooter spawner into Room

Finally fully functioning!!!

HN67 committed:
Cleanup warnings and nerf enemies

rupertraphael commented:
I'd merge and perhaps make players shoot diagonally or prevent the shooting enemies from shooting diagonally, and other gameplay balancing later?

rupertraphael commented on src/jerra/room/TextRoom.java:
```diff
@@ -3,124 +3,174 @@
+        // Adding to entities allows bulk updating
+        this.spawnEntity(player);
         // Registers the player as a spawner (for shooting)
         this.spawners.add(player);
```
Perhaps, use `room.spawnSpawner` for this?

HN67 committed:
Delegate player spawning to methods

HN67 committed:
Refactor direction storage and make enemies shoot orthogonally

HN67 committed:
Implement team system

Consider making a Gun have a team? and setting its fired bullets team to the same?

HN67 committed:
Experiment with smaller bullet
